### PR TITLE
Unicode support

### PR DESCRIPTION
**Pull request by jackvreeken**

### CODE REVIEW

**Comment by jackvreeken** [comment]
Adds support for unicode for Python 3.3+. See #2 for discussion.

File: src/frob.cc [hunk 1]
```diff
@@ -90,6 +89,96 @@ unsigned long ByteData(unsigned long addr) {
...
+
+  const long str_offset = unicode->state.ascii ? sizeof(PyASCIIObject)
+                                               : sizeof(PyCompactUnicodeObject);
+
+  // NOTE: From CPython commit c47adb04 onwards the kind matches directly to
+  // character size. This is different from the unicode format specification
+  // outlined in PEP 393, which still had only two bits allocated to the kind
+  // field.
+  const long ch_size = unicode->state.kind;
+  const long str_length = ch_size * unicode->length;
```
Comment (eklitzke) [inline src/frob.cc]: nit: this type should be `ssize_t`

File: src/frob.cc [hunk 2]
```diff
@@ -90,6 +89,96 @@ unsigned long ByteData(unsigned long addr) {
...
+  assert(unicode->state.compact);
+
+  const long str_offset = unicode->state.ascii ? sizeof(PyASCIIObject)
+                                               : sizeof(PyCompactUnicodeObject);
+
+  // NOTE: From CPython commit c47adb04 onwards the kind matches directly to
+  // character size. This is different from the unicode format specification
+  // outlined in PEP 393, which still had only two bits allocated to the kind
+  // field.
+  const long ch_size = unicode->state.kind;
```
Comment (eklitzke) [inline src/frob.cc]: nit: this should be `unsigned int`

File: src/frob.cc [hunk 3]
```diff
@@ -90,6 +89,96 @@ unsigned long ByteData(unsigned long addr) {
 static_assert(false, "uh oh, bad PYFLAME_PY_VERSION");
 #endif
 
+#if PYFLAME_PY_VERSION >= 34
+std::string StringDataPython3(pid_t pid, unsigned long addr) {
+  // TODO: This function only works for Python >= 3.3. Is it also possible to
+  // support older versions of Python 3?
+
+  // TODO: Can we guarantee that the same padding is used for the bitfield?
```
Comment (eklitzke) [inline src/frob.cc]: It *should* work if Python and Pyflame are compiled using the same compiler .This is a good point though.

File: src/frob.cc [hunk 4]
```diff
@@ -90,6 +89,96 @@ unsigned long ByteData(unsigned long addr) {
...
+    } else if (ch < 0x0800) {
+      /* Encode Latin-1 */
+      dump << (char)(0xc0 | (ch >> 6));
+      dump << (char)(0x80 | (ch & 0x3f));
+    } else if (ch < 0x10000) {
+      dump << (char)(0xe0 | (ch >> 12));
+      dump << (char)(0x80 | ((ch >> 6) & 0x3f));
+      dump << (char)(0x80 | (ch & 0x3f));
+    } else /* ch >= 0x10000 */
+    {
```
Comment (eklitzke) [inline src/frob.cc]: Can you move this brace up a line? You can still keep the inline comment on line 166. If you run the code through `clang-format` it should format things properly, assuming it doesn't completely butcher the bit twiddling stuff here.

File: tests/test_end_to_end.py [hunk 5]
```diff
@@ -212,6 +219,30 @@ def test_exclude_idle(sleeper):
         assert_flamegraph(line)
 
 
+if sys.version_info >= (3,3):
```
Comment (eklitzke) [inline tests/test_end_to_end.py]: You should use the pytest `skipif` feature here, the syntax is like:

```python
@pytest.mark.skipif(sys.version_info < (3, 3), reason="requires python3.3")
def test_function():
    ...
```

File: src/frob.cc (hunk 4 above)
Comment (jackvreeken) [inline src/frob.cc]: Oddly enough, clang-format did not change this line.

File: src/frob.cc (hunk 3 above)
Comment (jackvreeken) [inline src/frob.cc]: Do you want to remove the TODO, or keep it in?

**Pull request merged**